django-s3-file-field-client/src/client.ts

```diff
@@ -25,11 +31,51 @@ export interface UploadResult {
   state: 'aborted' | 'successful' | 'error';
 }
 
-export default class S3FFClient {
-  protected baseUrl: string;
+class Progressable {
+  private overheadPercentage = 0.0;
+  private mainPercentageSegments: number[] = [];
+  // This value was picked arbitrarily
+  private readonly overheadProportion = 0.20;
+
+  constructor(
+    private readonly onProgress?: ProgressCallback,
+  ) {}
+
+  private updateProgress(): void {
+    const mainPercentage = this.mainPercentageSegments
+      .reduce((total, value) => total + value, 0.0);
+    const percentage = (this.overheadPercentage * this.overheadProportion) + (mainPercentage * (1 - this.overheadProportion));
 
-  constructor(baseUrl: string) {
-    this.baseUrl = baseUrl;
+    if (this.onProgress) {
+      // Don't leak the "this" context out to the callback
+      this.onProgress.call(undefined, {
+        percentage,
+      });
+    }
+  }
+
+  protected updateOverheadProgress(percentage: number) {
+    this.overheadPercentage = percentage;
+    this.updateProgress();
+  }
+
+  protected initializeMainProgress(segments: number) {
+    this.mainPercentageSegments = new Array(segments).fill(0);
```

Comment:
I wrote up (but haven't yet tested) an alternative implementation that uses less memory. It leverages closures to track the progress incrementally:

```javascript
protected async uploadPart(chunk: ArrayBuffer, part: PartInfo, incrementProgress: Function): Promise<UploadedPart> {
    let lastLoaded = 0;
    const response = await axios.put(part.upload_url, chunk, {
      onUploadProgress: (e) => {
        const increment: number = e.loaded - lastLoaded;
        lastLoaded = e.loaded;
        incrementProgress(increment);
      },
    });
    const etag = response.headers.etag;
    return {
      part_number: part.part_number,
      size: part.size,
      etag,
    };
  }
```

...

```javascript
protected async uploadParts(file: File, parts: PartInfo[]): Promise<UploadedPart[]> {
    const buffer = await file.arrayBuffer();
    // indices track where in the buffer each part begins
    let index = 0;
    const indices: number[] = [];
    for (const part of parts) {
      indices.push(index);
      index += part.size;
    }

    let totalLoaded = 0;
    const incrementProgress = (increment: number) => {
      totalLoaded += increment;
      this.progressCallback({
        loaded: totalLoaded,
        total: file.size,
        lengthComputable: true,
      });
    };

    // upload each part of the buffer in parallel using the calculated indices
    return await Promise.all(parts.map(async (part, i) => {
      const chunk = buffer.slice(indices[i], indices[i] + part.size);
      return await this.uploadPart(chunk, part, incrementProgress);
    }));
  }
```

What do you think?